pkg/controller/deployment/deployment_controller.go

```diff
@@ -438,41 +438,150 @@ func (dc *DeploymentController) syncDeployment(key string) error {
+// are paused.
+func (dc *DeploymentController) sync(deployment *extensions.Deployment) error {
 	newRS, oldRSs, err := dc.getAllReplicaSetsAndSyncRevision(deployment, false)
 	if err != nil {
 		return err
 	}
-	allRSs := append(controller.FilterActiveReplicaSets(oldRSs), newRS)
 
-	// Sync deployment status
+	dc.scale(deployment, newRS, oldRSs)
+	dc.cleanupOldReplicaSets(oldRSs, deployment)
+
+	allRSs := append(oldRSs, newRS)
 	return dc.syncDeploymentStatus(allRSs, newRS, deployment)
 }
 
+// scale scales proportionally in order to mitigate risk. Otherwise, scaling up can increase the size
+// of the new replica set and scaling down can decrease the sizes of the old ones, both of which would
+// have the effect of hastening the rollout progress, which could produce a higher proportion of unavailable
+// replicas in the event of a problem with the rolled out template. Should run only on scaling events or
+// when a deployment is paused and not during the normal rollout process.
+func (dc *DeploymentController) scale(deployment *extensions.Deployment, newRS *extensions.ReplicaSet, oldRSs []*extensions.ReplicaSet) {
+	allRSs := oldRSs
+	if newRS != nil {
+		allRSs = append(allRSs, newRS)
+	}
+
+	// Paused and recreate deployments may not have a new replica set. Additionally,
+	// in case no old replica sets have pods, the controller should scale the new
+	// replica set to the full count.
+	if newRS != nil && len(controller.FilterActiveReplicaSets(oldRSs)) == 0 {
+		if newRS.Spec.Replicas == deployment.Spec.Replicas {
+			return
+		}
+		if newRS.Annotations == nil {
+			newRS.Annotations = make(map[string]string)
+		}
+		newRS.Annotations[deploymentutil.DesiredReplicasAnnotation] = fmt.Sprintf("%d", deployment.Spec.Replicas)
+		newRS.Annotations[deploymentutil.TotalReplicasAnnotation] = fmt.Sprintf("%d", deployment.Spec.Replicas)
+		glog.V(4).Infof("Scaling new replica set %q from %d to %d", newRS.Name, newRS.Spec.Replicas, deployment.Spec.Replicas)
+		if _, _, err := dc.scaleReplicaSetAndRecordEvent(newRS, deployment.Spec.Replicas, deployment); err != nil {
+			glog.Infof("Cannot scale replica set %q to %d replicas: %v", newRS.Name, deployment.Spec.Replicas, err)
+		}
+		return
+	}
+
+	// If the new replica set is saturated, old replica sets should be fully scaled down.
+	// This case handles replica set adoption during a saturated new replica set.
+	if deploymentutil.IsSaturated(deployment, newRS) {
+		for _, old := range controller.FilterActiveReplicaSets(oldRSs) {
+			if old.Annotations == nil {
+				old.Annotations = make(map[string]string)
+			}
+			old.Annotations[deploymentutil.DesiredReplicasAnnotation] = fmt.Sprintf("%d", deployment.Spec.Replicas)
+			old.Annotations[deploymentutil.TotalReplicasAnnotation] = fmt.Sprintf("%d", deployment.Spec.Replicas)
+			glog.V(4).Infof("Scaling down replica set %q from %d to 0", old.Name, old.Spec.Replicas)
+			if _, _, err := dc.scaleReplicaSetAndRecordEvent(old, 0, deployment); err != nil {
```

Comment:
What happens when this fails?  This call updates the argument replica field even if the update call fails, so it looks like the RS is updated in the deployment status, but it really isn't.  Since we can't perform all of the needed updates transactionally we need to sanely handle partial successes.